partner_affiliate/models/res_partner.py

```diff
@@ -21,3 +23,28 @@ class ResPartner(models.Model):
                                     string='Affiliates',
                                     domain=[('active', '=', True),
                                             ('is_company', '=', True)])
+
+    def get_original_address(self):
+        def convert(value):
+            return value.id if isinstance(value, models.BaseModel) else value
+
+        result = {'value': {key: convert(self[key])
+                            for key in self._address_fields()}}
+
+        return result
+
+    @api.onchange('parent_id')
+    def onchange_parent_id(self):
+        # Keep the original address info to set it back if its a company.
+        original_address = self.get_original_address()
+
+        new_partner = super(ResPartner, self).onchange_parent_id()
+
+        # When the affiliate is a company, we must set back its address
+        # because the super call changes its address by the new parent address.
+        # In addition, the type must be set to affiliate instead of contact.
+        if self.is_company:
```

Comment:
I think it is better to use `if new_partner and self.is_company:` condition, because `onchange_parent_id` may return `None`